

```
# This is formatted as code
```

# DMA Fall 20

**Note** : This entire lab will be manually evaluated.

Name : 'Annie Dang'
Collaborator : ''

# Lab 4: Neural Networks #


In [67]:
import pandas as pd
from sklearn.neural_network import MLPClassifier
from sklearn.svm import SVC

from sklearn.preprocessing import StandardScaler, MinMaxScaler
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
from sklearn.feature_extraction import DictVectorizer

from sklearn.pipeline import Pipeline
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split
from sklearn.model_selection import GridSearchCV, ParameterGrid

import numpy as np

import warnings
warnings.filterwarnings("ignore")

In [68]:
!wget http://people.ischool.berkeley.edu/~zp/course_datasets/lab_4_training.csv
!wget http://people.ischool.berkeley.edu/~zp/course_datasets/lab_4_test.csv

df_train = pd.read_csv('./lab_4_training.csv')
df_test = pd.read_csv('./lab_4_test.csv')
df_train.head()

URL transformed to HTTPS due to an HSTS policy
--2020-09-23 02:59:47--  https://people.ischool.berkeley.edu/~zp/course_datasets/lab_4_training.csv
Resolving people.ischool.berkeley.edu (people.ischool.berkeley.edu)... 128.32.78.16
Connecting to people.ischool.berkeley.edu (people.ischool.berkeley.edu)|128.32.78.16|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 105581 (103K) [text/csv]
Saving to: ‘lab_4_training.csv.3’

lab_4_training.csv. 100%[===================>] 103.11K   488KB/s    in 0.2s    

2020-09-23 02:59:48 (488 KB/s) - ‘lab_4_training.csv.3’ saved [105581/105581]

URL transformed to HTTPS due to an HSTS policy
--2020-09-23 02:59:48--  https://people.ischool.berkeley.edu/~zp/course_datasets/lab_4_test.csv
Resolving people.ischool.berkeley.edu (people.ischool.berkeley.edu)... 128.32.78.16
Connecting to people.ischool.berkeley.edu (people.ischool.berkeley.edu)|128.32.78.16|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 2652

Unnamed: 0 gender  age    year  ... exercisehours  musiccds  playgames  watchtv
0        1303   male   20  second  ...           5.0      50.0        1.0     15.0
1          36   male   20   third  ...           4.0      10.0        0.0      1.0
2         489   male   22  fourth  ...           2.0     150.0        1.0     10.0
3        1415   male   19  second  ...           5.0     100.0        0.0      7.0
4         616   male   22  fourth  ...           7.0      10.0        0.0      5.0

[5 rows x 15 columns]

***
### Question 1###
Calculate a baseline accuracy measure using the majority class, assuming a target variable of 'gender'. The majority class is the most common value of the target variable in a particular dataset. Accuracy is calculated as (true positives + true negatives) / (all negatives and positives)

**Question 1.a**  
Find the majority class in the training set. If you always predicted this class in the training set, what would your accuracy be?

In [69]:
df_train.groupby("gender").size()
TP = 855
FP = 735
FN = TN = 0
accuracy = (TP+TN)/(TP+FP+FN+TN)
accuracy

0.5377358490566038

###ANSWER: The accuracy would be 0.5377358490566038

**Question 1.b**   
If you always predicted this same class (majority from the training set) in the test set, what would your accuracy be?

In [70]:
df_test.groupby("gender").size()
TP = 208
FP = 190
FN = TN = 0
accuracy = (TP+TN)/(TP+FP+FN+TN)
accuracy

0.5226130653266332

###ANSWER: The accuracy would be 0.5226130653266332.

***
### Question 2 ###
Get started with Neural Networks.

   
Choose a NN implementation (eg: scikit-learn) and specify which you choose. Be sure the implementation allows you to modify the number of hidden layers and hidden nodes per layer.  

NOTE: When possible, specify the logsig (sigmoid/logistc) function as the transfer function (another word for activation function) and use Levenberg-Marquardt backpropagation (lbfgs). It is possible to specify logistic in Sklearn MLPclassifier (Neural net).  

**Question 2.a**   
Train a neural network with a single 10 node hidden layer. Only use the Height feature of the dataset to predict the Gender. You will have to change Gender to a 0 and 1 class. After training, use your trained model to predict the class using the height feature from the training set. What was the accuracy of this prediction?

In [71]:
import sklearn
from sklearn.tree import DecisionTreeClassifier
from sklearn.preprocessing import LabelEncoder

from sklearn.neural_network import MLPClassifier

lb_make = LabelEncoder()
df_train["gender"] = lb_make.fit_transform(df_train["gender"])

 
X = np.array(df_train["height"]).reshape(-1, 1) 
Y = df_train["gender"]
model = MLPClassifier(solver='lbfgs', hidden_layer_sizes=10, random_state=42,activation="logistic")
model.fit(X, Y)
model.predict(X)
model.score(X, Y)


0.8465408805031447

###ANSWER: The accuracy is 0.8465408805031447.

**Question 2.b**  
Take the trained model from question 2.a and use it to predict the test set. This can be accomplished by taking the trained model and giving it the Height feature values from the test set. What is the accuracy of this model on the test set?

In [72]:
df_test["gender"] = lb_make.fit_transform(df_test["gender"])

model.predict(np.array(df_test["height"]).reshape(-1, 1))
model.score(np.array(df_test["height"]).reshape(-1, 1) , df_test["gender"])

0.8542713567839196

###ANSWER: The accuracy of this model on the test set is 0.8542713567839196.

**Question 2.c**   
Neural Networks tend to prefer smaller, normalized feature values. Try taking the log of the height feature in both training and testing sets or use a Standard Scalar operation in SKlearn to centre and normalize the data between 0-1 for continuous values. Repeat question 2.a and 2.b with the log version or the normalized and centered version of this feature

In [73]:
model = MLPClassifier(solver='lbfgs', hidden_layer_sizes=10, random_state=42,activation="logistic")
model.fit(np.log(np.array(df_train["height"]).reshape(-1, 1)), df_train["gender"])

print(model.score(np.log(np.array(df_train["height"]).reshape(-1, 1)), df_train["gender"]))
print(model.score(np.log(np.array(df_test["height"]).reshape(-1, 1)), df_test["gender"]))

0.8465408805031447
0.8542713567839196


###ANSWER: The model acheived 0.8465408805031447 accuracy for the training set and 0.8542713567839196 for the test set.

*  0.8465408805031447 accuracy for the training set
*  0.8542713567839196 accuracy for the test set



***

### Question 3###
The rest of features in this dataset barring a few are categorical. No ML method accepts categorical features, so transform year, eyecolor, exercise into a set of binary features, one feature per unique original feature value, and mark the binary feature as ‘1’ if the feature value matches the original value and ‘0’ otherwise. Using only these binary variable transformed features, train and predict the class of the test set. What was your accuracy using Neural Network with a single 10 node hidden layer? During training, use a maximum number of iterations of 50.

In [74]:
df_train2 = df_train[['year', 'eyecolor', 'exercise', 'gender']]
df_train2 = pd.get_dummies(df_train2, columns=["year", "eyecolor", "exercise"])
df_test2 = df_test[['year', 'eyecolor', 'exercise', 'gender']]
df_test2 = pd.get_dummies(df_test2, columns=["year", "eyecolor", "exercise"])
X = df_train2.iloc[:, 1:]
Y = df_train2[['gender']]
X_test = df_test2.iloc[:, 1:]
Y_test = df_test2[['gender']]

model = MLPClassifier(solver='lbfgs', hidden_layer_sizes=10, random_state=42, activation="logistic", max_iter=50)
model.fit(X, Y)
print(model.score(X, Y))
print(model.score(X_test, Y_test))

0.5710691823899371
0.5728643216080402


###ANSWER: Accuracy for the test set is 0.5728643216080402.

***
### Question 4###
Using a NN, report the accuracy on  the test set of a model that trained only on the height and the eye color features of instances in the training set.

**Question 4.a**  
What is the accuracy on the test set using the original height values (no pre-processing) and eye color as a one-hot?

In [75]:
df_train3 = df_train[['gender', 'height', 'eyecolor']]
df_train3 = pd.get_dummies(df_train3, columns= ["eyecolor"])
X_train3 = df_train3.iloc[:, 1:]
Y_train3 = df_train3['gender']

df_test3 = df_test[['gender', 'height', 'eyecolor']]
df_test3 = pd.get_dummies(df_test3, columns = ['eyecolor'])
X_test3 = df_test3.iloc[:,1:]
Y_test3 = df_test3['gender']

model = MLPClassifier(solver='lbfgs', hidden_layer_sizes=10, random_state=42, activation="logistic", max_iter=50)
model.fit(X_train3, Y_train3)
print(model.score(X_train3, Y_train3))
print(model.score(X_test3, Y_test3))


0.7327044025157232
0.7361809045226131


###ANSWER: Accuracy for the test set using original height and eyecolor is 0.7361809045226131.

**Question 4.b**  
What is the accuracy on the test set using the log of height values (applied to both training and testing sets) and eye color as a one-hot?

In [76]:
df_train4 = df_train3
df_train4['height'] = np.log(df_train['height'])
X_train4 = df_train4.iloc[:, 1:]
Y_train4 = df_train4['gender']

df_test4 = df_test3
df_test4['height'] = np.log(df_test['height'])
X_test4 = df_test4.iloc[:,1:]
Y_test4 = df_test4['gender']
print(X_train4[['height']])
print(X_test4[['height']])
model = MLPClassifier(solver='lbfgs', hidden_layer_sizes=10, random_state=42, activation="logistic", max_iter=50)
model.fit(X_train4, Y_train4)
print(model.score(X_train4, Y_train4))
print(model.score(X_test4, Y_test4))

        height
0     4.290459
1     4.262680
2     4.317488
3     4.276666
4     4.262680
...        ...
1585  4.189655
1586  4.304065
1587  4.204693
1588  4.189655
1589  4.219508

[1590 rows x 1 columns]
       height
0    4.290459
1    4.262680
2    4.317488
3    4.276666
4    4.262680
..        ...
393  4.204693
394  4.158883
395  4.290459
396  4.127134
397  4.204693

[398 rows x 1 columns]
0.809433962264151
0.8040201005025126


###ANSWER: The accuracy of the test set using log height and eyecolor is 0.8040201005025126.

**Question 4.c**  
What is the accuracy on the test set using the Z-score of height values and eye color as a one-hot? 

Z-score is a normalization function. It is the value of a feature minus the average value for that feature (in the training set), divided by the standard deviation of that feature (in the training set). Remember that, whenever applying a function to a feature in the training set, it also has to be applied to that same feature in the test set.

In [77]:
def z_score(col):
  me = np.mean(col)
  std = np.std(col)
  return (col - me)/std

df_train5 = df_train3
df_train5['height'] = z_score(df_train['height'])
X_train5 = df_train5.iloc[:, 1:]
Y_train5 = df_train5['gender']

df_test5 = df_test3
df_test5['height'] = z_score(df_test['height'])
X_test5 = df_test5.iloc[:,1:]
Y_test5 = df_test5['gender']
print(X_train5[['height']])
print(X_test5[['height']])
model = MLPClassifier(solver='lbfgs', hidden_layer_sizes=10, random_state=42, activation="logistic", max_iter=50)
model.fit(X_train5, Y_train5)
print(model.score(X_train5, Y_train5))
print(model.score(X_test5, Y_test5))


        height
0     1.097386
1     0.626888
2     1.567885
3     0.862137
4     0.626888
...        ...
1585 -0.549359
1586  1.332635
1587 -0.314110
1588 -0.549359
1589 -0.078860

[1590 rows x 1 columns]
       height
0    1.100168
1    0.628837
2    1.571500
3    0.864502
4    0.628837
..        ...
393 -0.313826
394 -1.020823
395  1.100168
396 -1.492155
397 -0.313826

[398 rows x 1 columns]
0.8534591194968554
0.8693467336683417


###ANSWER: The accuracy for the testing set using the z-score of height and eyecolor is 0.8693467336683417.

***
### Question 5 ###
Repeat question 4 for exercise hours + eye color

In [78]:
from scipy import stats

df_train6 = df_train[['gender', 'exercisehours', 'eyecolor']]
df_train6 = pd.get_dummies(df_train6, columns= ["eyecolor"])
X_train6 = df_train6.iloc[:, 1:]
Y_train6 = df_train6['gender']

df_test6 = df_test[['gender', 'exercisehours', 'eyecolor']]
df_test6 = pd.get_dummies(df_test6, columns = ['eyecolor'])
X_test6 = df_test6.iloc[:,1:]
Y_test6 = df_test6['gender']

model = MLPClassifier(solver='lbfgs', hidden_layer_sizes=10, random_state=42, activation="logistic", max_iter=50)
model.fit(X_train6, Y_train6)
print(model.score(X_train6, Y_train6))
print("Accuracy of test using exercisehours & eyecolor: ",model.score(X_test6, Y_test6))

df_train7 = df_train[['gender', 'exercisehours', 'eyecolor']]
df_train7 = pd.get_dummies(df_train7, columns= ["eyecolor"])
df_train7['exercisehours'] = np.log(df_train['exercisehours'])
df_train7['exercisehours'] = df_train7['exercisehours'].replace(-np.inf, 0)
X_train7 = df_train7.iloc[:, 1:]
Y_train7 = df_train7['gender']

df_test7 = df_test[['gender', 'exercisehours', 'eyecolor']]
df_test7 = pd.get_dummies(df_test7, columns = ['eyecolor'])
df_test7['exercisehours'] = np.log(df_test['exercisehours'])
df_test7['exercisehours'] = df_test7['exercisehours'].replace(-np.inf, 0)
X_test7 = df_test7.iloc[:,1:]
Y_test7 = df_test7['gender']

model = MLPClassifier(solver='lbfgs', hidden_layer_sizes=10, random_state=42, activation="logistic", max_iter=50)
model.fit(X_train7, Y_train7)
print(model.score(X_train7, Y_train7))
print("Accuracy of test using log exercisehours & eyecolor: ",model.score(X_test7, Y_test7))

df_train8 = df_train[['gender', 'exercisehours', 'eyecolor']]
df_train8 = pd.get_dummies(df_train8, columns= ["eyecolor"])
df_train8['exercisehours'] = stats.zscore(df_train['exercisehours'])
X_train8 = df_train8.iloc[:, 1:]
Y_train8 = df_train8['gender']

df_test8 = df_test[['gender', 'exercisehours', 'eyecolor']]
df_test8 = pd.get_dummies(df_test8, columns = ['eyecolor'])
df_test8['exercisehours'] = stats.zscore(df_test['exercisehours'])
X_test8 = df_test8.iloc[:,1:]
Y_test8 = df_test8['gender']

model = MLPClassifier(solver='lbfgs', hidden_layer_sizes=10, random_state=42, activation="logistic", max_iter=50)
model.fit(X_train8, Y_train8)
model.predict(X_train8)
print(model.score(X_train8, Y_train8))
print("Accuracy of test using z-score exercisehours & eyecolor: ", model.score(X_test8, Y_test8))



0.5886792452830188
Accuracy of test using exercisehours & eyecolor:  0.5778894472361809
0.5786163522012578
Accuracy of test using log exercisehours & eyecolor:  0.5653266331658291
0.5786163522012578
Accuracy of test using z-score exercisehours & eyecolor:  0.5653266331658291


###ANSWER: The accuracy for test set using original exercise hours and eyecolor is 0.5778894472361809. The accuracy for using log exercise hours and eyecolor is 0.5653266331658291. The accuracy for z-score exercise hours and eyecolor is 0.5653266331658291.

***
### Question 6###
Combine the features from question 3, 4, and 5 (year, eyecolor, exercise, height, exercise hours). For numeric features use the best normalization method from questions 4 and 5.

**Question 6.a**  
What was the NN accuracy on the test set using the single 10 node hidden layer?

In [79]:
df_train9 = df_train[['gender', 'year', 'eyecolor', 'exercise', 'height', 'exercisehours']]
df_train9 = pd.get_dummies(df_train9, columns=['year', 'eyecolor', 'exercise'])
df_train9['height'] = z_score(df_train['height'])
df_train9['exercisehours'] = z_score(df_train['exercisehours'])
X_train9 = df_train9.iloc[:, 1:]
Y_train9 = df_train9['gender']

df_test9 = df_test[['gender', 'year', 'eyecolor', 'exercise', 'height', 'exercisehours']]
df_test9 = pd.get_dummies(df_test9, columns=['year', 'eyecolor', 'exercise'])
df_test9['height'] = z_score(df_test['height'])
df_test9['exercisehours'] = z_score(df_test['exercisehours'])
X_test9 = df_test9.iloc[:, 1:]
Y_test9 =  df_test9['gender']

model = MLPClassifier(solver='lbfgs', hidden_layer_sizes=10, random_state=42, activation="logistic", max_iter=50)
model.fit(X_train9, Y_train9)
model.predict(X_train9)
print(model.score(X_train9, Y_train9))
print("Accuracy of test using all features: ", model.score(X_test9, Y_test9))




0.8547169811320755
Accuracy of test using all features:  0.864321608040201


###ANSWER: The accuracy using the zscore of height/exercise hours, eyecolor, year, and exercise of the testing set is 0.864321608040201.

***
### Question 7- Bonus (10%)###
Can you improve your test set prediction accuracy by 5% or more?  

See how close to that milestone of improvement you can get by modifying the tuning parameters of  Neural Networks (the number of hidden layers, number of hidden nodes in each layer, the learning rate aka mu). A great guide to tuning parameters is explained in this guide: http://www.csie.ntu.edu.tw/~cjlin/papers/guide/guide.pdf. 

While the guide is specific to SVM and in particular the C and gamma parameters of the RBF kernel, the method applies to generally to any ML technique with tuning parameters.

Please also write a paragraph in a markdown cell below with an explanation of your approach and evaluation metrics.


In [80]:
df_train9 = df_train[['gender', 'year', 'eyecolor', 'exercise', 'height', 'exercisehours']]
df_train9 = pd.get_dummies(df_train9, columns=['year', 'eyecolor', 'exercise'])
df_train9['height'] = z_score(df_train['height'])
df_train9['exercisehours'] = z_score(df_train['exercisehours'])
X_train9 = df_train9.iloc[:, 1:]
Y_train9 = df_train9['gender']

df_test9 = df_test[['gender', 'year', 'eyecolor', 'exercise', 'height', 'exercisehours']]
df_test9 = pd.get_dummies(df_test9, columns=['year', 'eyecolor', 'exercise'])
df_test9['height'] = z_score(df_test['height'])
df_test9['exercisehours'] = z_score(df_test['exercisehours'])
X_test9 = df_test9.iloc[:, 1:]
Y_test9 =  df_test9['gender']

model = MLPClassifier(solver='lbfgs', hidden_layer_sizes=9, random_state=42, activation="logistic", alpha=1e-5)
model.fit(X_train9, Y_train9)
model.predict(X_train9)
print(model.score(X_train9, Y_train9))
print("Accuracy of test using all features: ", model.score(X_test9, Y_test9))


0.879874213836478
Accuracy of test using all features:  0.9045226130653267


###ANSWER: I was able to achieve 90.5% accuracy by setting the learning rate to 1e-5, removing the max iterations count, and setting the hidden layer sizes to 9. Although it's not a exactly 5% improvement, I played around with a lot of numbers and parameters and this was the highest accuracy I was able to achieve.